README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# ctbi

<!-- badges: start -->
<!-- badges: end -->

**Please cite** the following companion paper if you’re using the
**ctbi** package: Ritter, F.: Technical note: A procedure to clean,
decompose, and aggregate time series, Hydrol. Earth Syst. Sci., 27,
349–361, <https://doi.org/10.5194/hess-27-349-2023>, 2023.

The goal of **ctbi** is to clean, decompose, impute and aggregate
univariate time series. **Ctbi** stands for *Cyclic/Trend decomposition
using Bin Interpolation* : the time series is divided into a sequence of
non-overlapping bins. The long-term trend is a linear interpolation of
the mean values between successive bins and the cyclic component is the
mean stack of detrended data within all bins. Outliers present in the
residuals are flagged using an enhanced Boxplot rule (called **Logbox**)
that is adapted to non-Gaussian data and keeps the type I error at
$\frac{0.1}{\sqrt{n}}$ % (percentage of erroneously flagged outliers).
Logbox replaces the original 1.5 constant with $A \times \log(n)+B+C/n$.
The variable *n* is the sample size, $C = 36$ corrects for biases
emerging in small samples, and *A* and *B* are automatically calculated
on a predictor of the maximum tail weight. The strength of the cyclic
pattern within each bin is quantified by a new metric, the Stacked
Cycles Index (SCI), with SCI \~ 0 associated with no cyclicity and SCI =
1 a perfectly cyclic signal.

## Installation

You can install the latest version of **ctbi** on CRAN.

## Example

``` r
library(ctbi)
example1 <- data.frame(year = 1700:1988,sunspot = as.numeric(sunspot.year))
example1[sample(1:289,30),'sunspot'] <- NA # contaminate data with missing values
example1[c(5,30,50),'sunspot'] <- c(-50,300,400) # contaminate data with outliers
example1 <- example1[-(70:100),]
bin.period <- 11 # aggregation performed every 11 years (the year is numeric here)
bin.side <- 1989 # to capture the last year, 1988, in a complete bin
bin.FUN <- 'mean'
bin.max.f.NA <- 0.2 # maximum of 20% of missing data per bin
ylim <- c(0,Inf) # negative values are impossible

list.main <- ctbi(example1,bin.period=bin.period,
                       bin.side=bin.side,bin.FUN=bin.FUN,
                       ylim=ylim,bin.max.f.NA=bin.max.f.NA)
data0.example1 <- list.main$data0 # cleaned raw dataset
data1.example1 <- list.main$data1 # aggregated dataset.
mean.cycle.example1 <- list.main$mean.cycle # this data set shows a moderate seasonality
summary.bin.example1 <- list.main$summary.bin # confirmed with SCI = 0.50
summary.outlier.example1 <- list.main$summary.outlier
```


NEWS.md
# ctbi (development version)

# ctbi 2.0.5 (2023/01/20)

* documentation updated
* the case of 3 < n < 8 has been removed from ctbi.outlier
* The companion paper has been published with final DOI and added to ctbi

# ctbi 2.0.4 (2022/11/18)

* Fixed a minor mistake concerning the way residuals are handled for limited range of possible values (ex. [0,+Inf[ instead of ]-Inf,+Inf[)
* The description of the functions has been updated
* A temporary DOI has been attached until the review of the original manuscript is finished

# ctbi 2.0.3 (2022/07/27)

* summary.bin has been added to the output of ctbi, and it condenses the information about the size of a bin (bin.size), the minimum number of points for a bin to be accepted (bin.size.min.accepted) and the Stacked Cycle Index (SCI)
* the mean of mean cycle has been subtracted and added to the long.term
* examples have been updated

# ctbi 2.0.2 (2022/07/25)

* summary.outlier has been added to the output of ctbi, and gives information about coeff.outlier (coefficients A,B and C), the lower/upper outlier threshold, the number of points used, the value of m.star.
* a minor mistake has been fixed for the residuals
* the outliers are now present in the column residuals to compare them to the lower/upper threshold

# ctbi 2.0.1 (2022/07/22)

* ctbi.outliers has been changed with ctbi.outlier. This function now only takes an input vector and can handle any univariate datasets (no necessarily based on residuals).
* The Logbox has been updated. g_A(m.star)=0.2294exp(2.9416m.star-0.0512m.star^2-0.0684m.star^3) and g_B(m.star)=1.0585+15.6960m.star-17.3618m.star^2+28.3511m.star^3-11.4726m.star^4 will respectively calculate the A and B coefficients with m.star = (q(0.875)-q(0.625))/(q(0.75)-q(0.25))-0.6165 for right-skewed distributions, bounded by [0,2]. The C coefficients is fixed to 50 (Pearson Family).
* A "gaussian" method for Logbox has been added, with coeff.outlier = c(0.08,2,33) 
* for n < 9, the MAD will be used instead the boxplot rule, with a multiplying factor of 10 (instead of usually 3) to limit the number of false positive to 0.5%.

# ctbi 2.0.0 (2022/06/10)

* The LogBox method has changed from alpha = 1+klog(n) to alpha = Alog(n)+B+C/n
* k.outliers has been replaced by coeffs.outlier = c(A,B,C).
* SCI.min = Inf has been replaced by SCI.min = NA when no values are imputed

# ctbi 1.0.1 (2022/03/08)

* F or T have been replaced with FALSE or TRUE
* Example 2 has been added to the help of ctbi
* Value has been added to the help of ctbi.plot and hidd.seq

# ctbi 1.0.0 (2022/03/01)

* ORCID of Francois Ritter has been updated.
* n.bin.min <- floor(bin.size\*(1-bin.max.f.NA)) has been replaced with n.bin.min <- ceiling(bin.size\*(1-bin.max.f.NA)). This means that for bin.max.f.NA = 0.2 (bins with at least 80% of data are accepted) and bin.size = 12 (e.g., 12 months of data per bin), a bin with 9 months of data will be rejected.
* n.bin.min (minimum number of data points for a bin to be accepted) has been added as an output of ctbi in list.main.
* The starting and ending boundaries of the long.term interpolation were treated as NA values. This has been removed.
* The numeric class of data0[,y] has been forced to avoid problems with integers.
* Fixed a minor problem with the error message concerning k.outliers
* Fixed a plotting issue with the y range.


R/ctbi.outlier.R
#' @title ctbi.outlier
#'
#' @description \bold{Please cite} the following companion paper if you're using the \code{ctbi} package: Ritter, F.: Technical note: A procedure to clean, decompose, and aggregate time series, Hydrol. Earth Syst. Sci., 27, 349–361, https://doi.org/10.5194/hess-27-349-2023, 2023.
#'
#' Outliers in an univariate dataset \code{y} are flagged using an enhanced box plot rule (called \bold{Logbox}, input: \code{coeff.outlier}) that is adapted to non-Gaussian data and keeps the type I error at \eqn{\frac{0.1}{\sqrt{n}}} \% (percentage of erroneously flagged outliers).
#'
#' The box plot rule flags data points as outliers if they are below \eqn{L} or above \eqn{U} using the sample quantile \eqn{q}:
#'
#' \eqn{L = q(0.25)-\alpha \times (q(0.75)- q(0.25))}
#'
#' \eqn{U = q(0.75)+\alpha \times (q(0.75)- q(0.25))}
#'
#' \bold{Logbox} replaces the original \eqn{\alpha = 1.5} constant of the box plot rule with \eqn{\alpha = A \times \log(n)+B+\frac{C}{n}}. The variable \eqn{n \geq 9} is the sample size, \eqn{C = 36} corrects biases emerging in small samples, and \eqn{A} and \eqn{B} are automatically calculated on a predictor of the maximum tail weight defined as \eqn{m_{*} = \max(m_{-},m_{+})-0.6165}.
#'
#' The two functions (\eqn{m_{-}},\eqn{m_{+}}) are defined as:
#'
#' \eqn{m_{-} = \frac{q(0.875)- q(0.625)}{q(0.75)- q(0.25)}}
#'
#' \eqn{m_{+} = \frac{q(0.375)- q(0.125)}{q(0.75)- q(0.25)}}
#'
#' And finally, \eqn{A = f_{A}(}\eqn{m_{*}}\eqn{)} and \eqn{B = f_{B}(}\eqn{m_{*}}\eqn{)} with \eqn{m_{*}} restricted to [0,2]. The functions \eqn{(f_{A},f_{B})} are defined as:
#'
#' \eqn{f_{A}(x) = 0.2294\exp(2.9416x-0.0512x^{2}-0.0684x^{3})}
#'
#' \eqn{f_{B}(x) = 1.0585+15.6960x-17.3618x^{2}+28.3511x^{3}-11.4726x^{4}}
#'
#' Both functions have been calibrated on the Generalized Extreme Value and Pearson families.
#'
#' @param y univariate data (numeric vector)
#' @param coeff.outlier one of \code{coeff.outlier} = 'auto' (default value), \code{coeff.outlier} = 'gaussian', \code{coeff.outlier} = c(A,B,C) or \code{coeff.outlier} = \code{NA}. If \code{coeff.outlier} = 'auto', C = 36 and the coefficients A and B are calculated on \eqn{m_{*}}. If \code{coeff.outlier} = 'gaussian', \code{coeff.outlier} = c(0.08,2,36), adapted to the Gaussian distribution. If \code{coeff.outlier} = \code{NA}, no outliers are flagged
#'
#' @return A list that contains:
#' @return \bold{xy}, a two columns data frame that contains the clean data (first column) and the outliers (second column)
#' @return \bold{summary.outlier}, a vector that contains A, B, C, \eqn{m_{*}}, the size of the residuals (n), and the lower and upper outlier threshold
#' @examples
#' x <- runif(30)
#' x[c(5,10,20)] <- c(-10,15,30)
#' example1 <- ctbi.outlier(x)
#' @export

ctbi.outlier <- function(y,coeff.outlier='auto')
{
  # type of outlier checking
  log.user.coeff <- FALSE
  log.no.outlier.checking <- FALSE
  if(length(coeff.outlier) == 1)
  {
    if(is.na(coeff.outlier))
    {
      log.no.outlier.checking <- TRUE
    }else
    {
      if(coeff.outlier == 'gaussian')
      {
        log.user.coeff <- TRUE
        coeff.outlier <- c(0.08,2,36) # gaussian values
      }
    }
  }else
  {
    if(length(coeff.outlier) == 3)
    {
      coeff.outlier[is.na(coeff.outlier)] <- 0
      log.user.coeff <- TRUE
    }else
    {
      log.no.outlier.checking <- TRUE
    }
  }

  xy <- data.frame(y.clean=y,outliers=rep(NA,length(y)),stringsAsFactors = F)

  if(log.no.outlier.checking)
  {
    A.coeff <- NA
    B.coeff <- NA
    C.coeff <- NA
    m.star <- NA
    n.residuals <- NA
    lower.outlier.threshold <- NA
    upper.outlier.threshold <- NA
  }else
  {
    res.nonylim <- as.numeric(xy[,'y.clean'])
    res.nonylim <- res.nonylim[!is.na(res.nonylim)]
    n.residuals <- length(res.nonylim)

    if(n.residuals > 8)
    {
      if(log.user.coeff)
      {
        q0.25 <- as.numeric(quantile(res.nonylim,0.25))
        q0.75 <- as.numeric(quantile(res.nonylim,0.75))
        A.coeff <- coeff.outlier[1]
        B.coeff <- coeff.outlier[2]
        C.coeff <- coeff.outlier[3]
        m.star <- NA
      }else
      {
        q0.125 <- as.numeric(quantile(res.nonylim,0.125))
        q0.25 <- as.numeric(quantile(res.nonylim,0.25))
        q0.375 <- as.numeric(quantile(res.nonylim,0.375))
        q0.625 <- as.numeric(quantile(res.nonylim,0.625))
        q0.75 <- as.numeric(quantile(res.nonylim,0.75))
        q0.875 <- as.numeric(quantile(res.nonylim,0.875))

        m.plus <- (q0.875-q0.625)/(q0.75-q0.25)
        m.minus <- (q0.375-q0.125)/(q0.75-q0.25)

        if(!is.na(m.minus) & !is.na(m.plus))
        {
          m.star <- max(c(m.plus,m.minus))-0.6165

          if(m.star < 0){m.star <- 0}
          if(m.star > 2){m.star <- 2}

          # version 13
          a1 <- 0.2294
          a2 <- 2.9416
          a3 <- -0.0512
          a4 <- -0.0684
          A.coeff <- round(a1*exp(a2*m.star+a3*m.star^2+a4*m.star^3),digits=2)

          # version 13
          b1 <- 1.0585
          b2 <- 15.6960
          b3 <- -17.3618
          b4 <- 28.3511
          b5 <- -11.4726
          B.coeff <- round(b1+b2*m.star+b3*(m.star^2)+b4*(m.star^3)+b5*(m.star^4),digits=2)

          C.coeff <- 36

        }else
        {
          m.star <- NA
          A.coeff <- NA
          B.coeff <- NA
          C.coeff <- NA
          lower.outlier.threshold <- NA
          upper.outlier.threshold <- NA
        }
      }


      if(!is.na(A.coeff))
      {
        alpha <- A.coeff*log(n.residuals)+B.coeff+C.coeff/n.residuals
        lower.outlier.threshold <- q0.25 - alpha*(q0.75-q0.25)
        upper.outlier.threshold <- q0.75 + alpha*(q0.75-q0.25)

        read_ <- (xy[,'y.clean'] < lower.outlier.threshold) | (xy[,'y.clean'] > upper.outlier.threshold)
        read_[is.na(read_)] <- FALSE
        read_ <- as.logical(read_)
        xy[read_,'outliers'] <- xy[read_,'y.clean']
        xy[read_,'y.clean'] <- NA_real_
      }
    }
    else
    {
      A.coeff <- NA
      B.coeff <- NA
      C.coeff <- NA
      m.star <- NA
      lower.outlier.threshold <- NA
      upper.outlier.threshold <- NA
    }

  }

  summary.xy <- c(A.coeff,B.coeff,C.coeff,m.star,n.residuals,lower.outlier.threshold,upper.outlier.threshold)
  names(summary.xy) <- c('A','B','C','m.star','n','lower.outlier.threshold','upper.outlier.threshold')
  list.out <- list(xy=xy,summary.outlier=summary.xy)
  return(list.out)
}
